### PR TITLE
[API][Homepage] Latest products

### PR DESCRIPTION
**Pull request by AdamKasp**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | part of #11250 
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/Context/RequestBasedChannelContext.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use Sylius\Bundle\ApiBundle\Metadata\Resource\OperationCheckerTrait;
+use Sylius\Component\Channel\Context\ChannelContextInterface;
+use Sylius\Component\Channel\Context\ChannelNotFoundException;
+use Sylius\Component\Channel\Model\ChannelInterface;
+use Sylius\Component\Channel\Repository\ChannelRepositoryInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+final class RequestBasedChannelContext implements ChannelContextInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Context/RequestBasedChannelContext.php]: This class is not needed. We will not provide a channel in the path attribute. Let's re-use what we have in core instead.

File: src/Sylius/Bundle/ApiBundle/Context/RequestBasedLocaleContext.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+namespace Sylius\Bundle\ApiBundle\Context;
+
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use Sylius\Bundle\ApiBundle\Metadata\Resource\OperationCheckerTrait;
+use Sylius\Component\Locale\Context\LocaleContextInterface;
+use Sylius\Component\Locale\Context\LocaleNotFoundException;
+use Sylius\Component\Locale\Provider\LocaleProviderInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+final class RequestBasedLocaleContext implements LocaleContextInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Context/RequestBasedLocaleContext.php]: For now, it is not needed. Let's re-use what we have in core instead.

File: src/Sylius/Bundle/ApiBundle/DataProvider/ProductCollectionDataProvider.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+use Webmozart\Assert\Assert;
+
+final class ProductCollectionDataProvider implements CollectionDataProviderInterface, RestrictedDataProviderInterface
+{
+    /** @var ProductRepositoryInterface */
+    private $productRepository;
+
+    public function supports(string $resourceClass, string $operationName = null, array $context = []): bool
+    {
+        return is_a($resourceClass, ProductInterface::class, true) && $operationName === "";
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/DataProvider/ProductCollectionDataProvider.php]:
```suggestion
        return is_a($resourceClass, ProductInterface::class, true) && $operationName === "product_latest";
```

File: src/Sylius/Bundle/ApiBundle/Metadata/Resource/AttributeKeys.php [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Metadata\Resource;
+
+final class AttributeKeys
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Metadata/Resource/AttributeKeys.php]: Not needed I suppose

File: src/Sylius/Bundle/ApiBundle/Metadata/Resource/OperationCheckerTrait.php [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Metadata\Resource;
+
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use ApiPlatform\Core\Util\RequestAttributesExtractor;
+use Symfony\Component\HttpFoundation\Request;
+
+trait OperationCheckerTrait
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Metadata/Resource/OperationCheckerTrait.php]: Probably not needed

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml [hunk 6]
```diff
@@ -34,11 +34,24 @@
...
                 </attribute>
             </collectionOperation>
+
             <collectionOperation name="post">
                 <attribute name="denormalization_context">
                     <attribute name="groups">product:create</attribute>
                 </attribute>
             </collectionOperation>
+
+            <collectionOperation name="get_latest">
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml]:
```suggestion
            <collectionOperation name="product_latest">
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml [hunk 7]
```diff
@@ -34,11 +34,24 @@
...
+                <attribute name="method">GET</attribute>
+                <attribute name="path">/{channelCode}/product-latest/</attribute>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml]:
```suggestion
                <attribute name="path">/product-latest/</attribute>
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml [hunk 8]
```diff
@@ -34,11 +34,24 @@
...
+                <attribute name="normalization_context">
+                    <attribute name="groups">
+                        <attribute>product:read</attribute>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml]:
```suggestion
                        <attribute>shop:product:read</attribute>
```
Probably all current attributes should be renamed to `admin:product:x`

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml [hunk 9]
```diff
@@ -34,11 +34,24 @@
...
+                        <attribute>product:index</attribute>
+                    </attribute>
+                </attribute>
+                <attribute name="sylius_api">true</attribute>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml]: Perhaps not needed

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml (hunk 6 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml]: At least should correspond to https://github.com/Sylius/Sylius/pull/11291/files#diff-2620115d9fd4902065fef0e4af355ceeR30

File: src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/ParametersContextBuilder.php [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace Sylius\Bundle\ApiBundle\SerializerContextBuilder;
+
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use ApiPlatform\Core\Serializer\SerializerContextBuilderInterface;
+use Sylius\Bundle\ApiBundle\Metadata\Resource\AttributeKeys;
+use Sylius\Bundle\ResourceBundle\Controller\ParametersParserInterface;
+use Symfony\Component\HttpFoundation\Request;
+
+final class ParametersContextBuilder implements SerializerContextBuilderInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/ParametersContextBuilder.php]: It is maybe a problem with my comprehension, but I would say that this class is not needed as well

File: src/Sylius/Bundle/ApiBundle/DataProvider/ProductCollectionDataProvider.php [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+     public function getCollection(string $resourceClass, string $operationName = null, array $context = []): iterable
+    {
+        Assert::keyExists($context, ContextKeys::CHANNEL);
+        Assert::keyExists($context, ContextKeys::LOCALE_CODE);
+
+        $channel = $context[ContextKeys::CHANNEL];
+        $locale = $context[ContextKeys::LOCALE_CODE];
+
+        return $this->productRepository->findLatestByChannel($channel, $locale, 3);
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/DataProvider/ProductCollectionDataProvider.php]: 3 could be parametrized

File: src/Sylius/Bundle/ApiBundle/DataProvider/ProductCollectionDataProvider.php (hunk 3 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/ApiBundle/DataProvider/ProductCollectionDataProvider.php]:
```suggestion
        return is_a($resourceClass, ProductInterface::class, true) && $operationName === "get_latest";
```

Comment (AdamKasp) [reply src/Sylius/Bundle/ApiBundle/DataProvider/ProductCollectionDataProvider.php]: okey, but for now i prefer name "get_latest"

File: src/Sylius/Behat/Client/ApiClientInterface.php [hunk 12]
```diff
@@ -40,6 +40,8 @@ public function applyTransition(string $id, string $transition, array $content =
 
     public function customItemAction(string $id, string $type, string $action): Response;
 
+    public function customAction(string $type, string $action): Response;
```
Comment (GSadee) [inline src/Sylius/Behat/Client/ApiClientInterface.php]:
```suggestion
    public function customAction(string $method, string $url): Response;
```
and these arguments could be in the reverse order imo

File: src/Sylius/Behat/Resources/config/suites.yml [hunk 13]
```diff
@@ -19,6 +19,7 @@ imports:
     - suites/api/taxon/managing_taxons.yml
     - suites/api/user/managing_administrators.yml
     - suites/api/user/managing_customer_groups.yml
+    - suites/api/homepage/viewing_products.yml
```
Comment (GSadee) [inline src/Sylius/Behat/Resources/config/suites.yml]: Wrong order of files

File: src/Sylius/Bundle/ApiBundle/DataProvider/ProductCollectionDataProvider.php [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+        return is_a($resourceClass, ProductInterface::class, true) && $operationName === "get_latest";
+    }
+
+     public function getCollection(string $resourceClass, string $operationName = null, array $context = []): iterable
+    {
+        Assert::keyExists($context, ContextKeys::CHANNEL);
+        Assert::keyExists($context, ContextKeys::LOCALE_CODE);
+
+        $channel = $context[ContextKeys::CHANNEL];
+        $locale = $context[ContextKeys::LOCALE_CODE];
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/DataProvider/ProductCollectionDataProvider.php]:
```suggestion
        $localeCode = $context[ContextKeys::LOCALE_CODE];
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Product.xml [hunk 15]
```diff
@@ -17,44 +17,55 @@
...
         <attribute name="reviews">
-            <group>product:read</group>
+            <group>admin:product:read</group>
+            <group>shop:product:read</group>
             <group>product:create</group>
         </attribute>
         <attribute name="options">
-            <group>product:read</group>
+            <group>admin:product:read</group>
+
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Product.xml]: ? 😄

File: src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/ChannelContextBuilder.php [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+
+use ApiPlatform\Core\Serializer\SerializerContextBuilderInterface;
+use Sylius\Bundle\ApiBundle\Serializer\ContextKeys;
+use Sylius\Component\Channel\Context\ChannelContextInterface;
+use Symfony\Component\HttpFoundation\Request;
+
+final class ChannelContextBuilder implements SerializerContextBuilderInterface
+{
+    private $decorated;
+    private $channelContext;
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/ChannelContextBuilder.php]: Add docblocks

File: src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/ChannelContextBuilder.php [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+
+    public function __construct(SerializerContextBuilderInterface $decorated, ChannelContextInterface $channelContext)
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/ChannelContextBuilder.php]:
```suggestion
    public function __construct(SerializerContextBuilderInterface $decoratedContextBuilder, ChannelContextInterface $channelContext)
```
?

File: features/homepage/viewing_latest_products.feature [hunk 18]
```diff
@@ -7,8 +7,10 @@ Feature: Viewing a latest product list
     Background:
         Given the store operates on a single channel in "United States"
         And this channel has "Belvedere Vodka", "Coconaut Liqeur", "Chopin Chocolate Liquer" and "Capitan Morgan White Rum" products
+#        temporarary, for testing purpose, product viewing should be on different path(without mandatory admin login)
+        And I am logged in as an administrator
 
-    @ui
+     @api
```
Comment (lchrusciel) [inline features/homepage/viewing_latest_products.feature]: Can we revert it?

File: src/Sylius/Behat/Client/ApiClientInterface.php (hunk 12 above)
Comment (lchrusciel) [reply src/Sylius/Behat/Client/ApiClientInterface.php]: However, URL could suggest that we are providing a full path here. Not sure, if action is not a better wording.

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml [hunk 19]
```diff
@@ -18,7 +18,8 @@
     <resource class="%sylius.model.product.class%" shortName="Product">
         <attribute name="normalization_context">
             <attribute name="groups">
-                <attribute>product:read</attribute>
+                <attribute>admin:product:read</attribute>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml]: If you didn't change it in all of the other places, let's leave admin serialization without the prefix
```suggestion
                <attribute>product:read</attribute>
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml [hunk 20]
```diff
@@ -34,11 +35,23 @@
...
                 </attribute>
             </collectionOperation>
+
+            <collectionOperation name="get_latest">
+                <attribute name="method">GET</attribute>
+                <attribute name="path">/latest/</attribute>
+                <attribute name="normalization_context">
+                    <attribute name="groups">
+                        <attribute>shop:product:read</attribute>
+                        <attribute>product:index</attribute>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml]: Why do we have two serialization groups here?

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml (hunk 19 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml]: We should introduce this prefix in separate PR

File: src/Sylius/Bundle/ApiBundle/Resources/config/services.xml [hunk 21]
```diff
@@ -65,5 +66,10 @@
         <service id="sylius.api.swagger_documentation_normalizer" class="Sylius\Bundle\ApiBundle\Swagger\DocumentationNormalizer" decorates="api_platform.swagger.normalizer.documentation" public="true" autoconfigure="false">
             <argument type="service" id="sylius.api.swagger_documentation_normalizer.inner" />
         </service>
+
+        <service id="sylius.api.collection_data_provider.product" class="Sylius\Bundle\ApiBundle\DataProvider\ProductCollectionDataProvider">
+            <argument type="service" id="sylius.repository.product" />
+            <tag name="api_platform.collection_data_provider" />
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services.xml]:
```suggestion
            <tag name="api_platform.collection_data_provider" priority=10 />
```

File: src/Sylius/Behat/Client/ApiClientInterface.php (hunk 12 above)
Comment (GSadee) [reply src/Sylius/Behat/Client/ApiClientInterface.php]: And in the implementation this argument is used as a full path

File: features/homepage/viewing_latest_products.feature [hunk 22]
```diff
@@ -7,13 +7,15 @@ Feature: Viewing a latest product list
     Background:
         Given the store operates on a single channel in "United States"
         And this channel has "Belvedere Vodka", "Coconaut Liqeur", "Chopin Chocolate Liquer" and "Capitan Morgan White Rum" products
+        # temporary solution, it should be remove after authorisation of new-api/product/latest done.
+        And I am logged in as an administrator
 
-    @ui
+     @ui @api
```
Comment (lchrusciel) [inline features/homepage/viewing_latest_products.feature]:
```suggestion
    @ui @api
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/contextBuilders.xml [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+
+<container xmlns="http://symfony.com/schema/dic/services"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <services>
+        <defaults public="false" />
+
+        <service id="sylius_api.serializer_context_builder.channel"
+                 class="Sylius\Bundle\ApiBundle\SerializerContextBuilder\ChannelContextBuilder"
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/contextBuilders.xml]: Wrong indent
```suggestion
             class="Sylius\Bundle\ApiBundle\SerializerContextBuilder\ChannelContextBuilder"
```
Am I correct? /cc @GSadee

File: src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/LocaleContextBuilder.php [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        $context = $this->decorated->createFromRequest($request, $normalization, $extractedAttributes);
+        try {
+            $context[ContextKeys::LOCALE_CODE] = $this->localeContext->getLocaleCode();
+            $context[ContextKeys::FALLBACK_LOCALE_CODE] = $this->translationLocaleProvider->getDefaultLocaleCode();
+        } catch (NotFoundHttpException $exception) {
+        }
+
+        return $context;
+    }
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/LocaleContextBuilder.php]:
```suggestion
    public function createFromRequest(Request $request, bool $normalization, ?array $extractedAttributes = null): array
    {
        $context = $this->decorated->createFromRequest($request, $normalization, $extractedAttributes);
        try {
            $context[ContextKeys::LOCALE_CODE] = $this->localeContext->getLocaleCode();
        } catch (LocaleNotFound $exception) {
        }

        return $context;
    }
```

File: src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/LocaleContextBuilder.php [hunk 25]
```diff
@@ -0,0 +1,47 @@
...
+        $this->translationLocaleProvider = $translationLocaleProvider;
+    }
+
+    public function createFromRequest(Request $request, bool $normalization, ?array $extractedAttributes = null): array
+    {
+        $context = $this->decorated->createFromRequest($request, $normalization, $extractedAttributes);
+        try {
+            $context[ContextKeys::LOCALE_CODE] = $this->localeContext->getLocaleCode();
+            $context[ContextKeys::FALLBACK_LOCALE_CODE] = $this->translationLocaleProvider->getDefaultLocaleCode();
+        } catch (NotFoundHttpException $exception) {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/LocaleContextBuilder.php]:
```suggestion
        } catch (LocaleNotFoundException $exception) {
```

File: src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/ChannelContextBuilder.php [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+        $this->channelContext = $channelContext;
+    }
+
+    public function createFromRequest(Request $request, bool $normalization, ?array $extractedAttributes = null): array
+    {
+        $context = $this->decoratedContextBuilder->createFromRequest($request, $normalization, $extractedAttributes);
+
+        try {
+            $context[ContextKeys::CHANNEL] = $this->channelContext->getChannel();
+
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/ChannelContextBuilder.php]:
```suggestion
```

File: src/Sylius/Bundle/ApiBundle/Serializer/ContextKeys.php [hunk 27]
```diff
@@ -0,0 +1,25 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Serializer;
+
+final class ContextKeys
+{
+    public const CHANNEL = 'sylius_api_channel';
+    public const LOCALE_CODE = 'sylius_api_locale_code';
+    public const FALLBACK_LOCALE_CODE = 'sylius_api_fallback_locale_code';
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Serializer/ContextKeys.php]:
```suggestion
```

File: src/Sylius/Behat/Context/Api/Shop/HomePageContext.php [hunk 28]
```diff
@@ -0,0 +1,59 @@
...
+        $this->responseChecker = $responseChecker;
+        $this->sharedStorage = $sharedStorage;
+    }
+
+    /**
+     * @When I check latest products
+     */
+    public function iCheckLatestProducts(): void
+    {
+        $this->productsClient->customAction('shop/products/latest', HttpRequest::METHOD_GET);
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Shop/HomePageContext.php]: Are you sure about this path?

```suggestion
        $this->productsClient->customAction('new-api/shop/products/latest', HttpRequest::METHOD_GET);
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml [hunk 29]
```diff
@@ -31,27 +29,45 @@
...
             <collectionOperation name="post">
+                <attribute name="path">admin/products</attribute>
                 <attribute name="denormalization_context">
                     <attribute name="groups">product:create</attribute>
                 </attribute>
             </collectionOperation>
+
+            <collectionOperation name="get_latest">
+                <attribute name="method">GET</attribute>
+                <attribute name="path">/shop/products/latest</attribute>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Product.xml]: To keep it consistent with admin paths definition
```suggestion
                <attribute name="path">shop/products/latest</attribute>
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Product.xml [hunk 30]
```diff
@@ -18,10 +18,12 @@
     <class name="Sylius\Component\Core\Model\Product">
         <attribute name="id">
             <group>product:read</group>
+            <group>shop:product:read</group>
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Product.xml]:
```suggestion
            <group>admin:product:read</group>
            <group>shop:product:read</group>
```
?

File: src/Sylius/Bundle/ApiBundle/Resources/config/services.xml [hunk 31]
```diff
@@ -13,7 +13,8 @@
 
 <container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
     <imports>
-        <import resource="services/filters.xml"/>
+        <import resource="services/filters.xml" />
+        <import resource="services/contextBuilders.xml" />
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services.xml]:
```suggestion
        <import resource="services/context_builders.xml" />
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/contextBuilders.xml [hunk 32]
```diff
@@ -0,0 +1,40 @@
...
+
+ For the full copyright and license information, please view the LICENSE
+ file that was distributed with this source code.
+
+-->
+
+<container xmlns="http://symfony.com/schema/dic/services"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/contextBuilders.xml]:
```suggestion
           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd"
>
```

File: src/Sylius/Bundle/ApiBundle/Serializer/ContextKeys.php [hunk 33]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Serializer;
+
+final class ContextKeys
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Serializer/ContextKeys.php]:
```suggestion
class ContextKeys
```

File: src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/ChannelContextBuilder.php [hunk 34]
```diff
@@ -0,0 +1,47 @@
...
+{
+    /** @var SerializerContextBuilderInterface  */
+    private $decoratedContextBuilder;
+
+    /** @var ChannelContextInterface  */
+    private $channelContext;
+
+    public function __construct(SerializerContextBuilderInterface $decorated, ChannelContextInterface $channelContext)
+    {
+        $this->decoratedContextBuilder = $decorated;
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/ChannelContextBuilder.php]:
```suggestion
        $this->decoratedContextBuilder = $decoratedContextBuilder;
```

File: src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/ChannelContextBuilder.php [hunk 35]
```diff
@@ -0,0 +1,47 @@
...
+        $this->channelContext = $channelContext;
+    }
+
+    public function createFromRequest(Request $request, bool $normalization, ?array $extractedAttributes = null): array
+    {
+        $context = $this->decoratedContextBuilder->createFromRequest($request, $normalization, $extractedAttributes);
+        try {
+            $context[ContextKeys::CHANNEL] = $this->channelContext->getChannel();
+
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/ChannelContextBuilder.php]: Unnecessary blank line

File: src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/LocaleContextBuilder.php [hunk 36]
```diff
@@ -0,0 +1,46 @@
...
+use Sylius\Component\Locale\Context\LocaleContextInterface;
+use Sylius\Component\Locale\Context\LocaleNotFoundException;
+use Sylius\Component\Resource\Translation\Provider\TranslationLocaleProviderInterface;
+use Symfony\Component\HttpFoundation\Request;
+
+final class LocaleContextBuilder implements SerializerContextBuilderInterface
+{
+    private $decorated;
+    private $localeContext;
+    private $translationLocaleProvider;
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/LocaleContextBuilder.php]: Add docblocks

File: src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/LocaleContextBuilder.php [hunk 37]
```diff
@@ -0,0 +1,46 @@
...
+
+    public function __construct(SerializerContextBuilderInterface $decorated, LocaleContextInterface $localeContext, TranslationLocaleProviderInterface $translationLocaleProvider)
+    {
+        $this->decorated = $decorated;
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/SerializerContextBuilder/LocaleContextBuilder.php]: Change `decorated` to `decoratedContextBuilder` the same as I've commented you in `ChannelContextBuilder`

File: src/Sylius/Bundle/ApiBundle/DataProvider/LatestProductCollectionDataProvider.php [hunk 38]
```diff
@@ -0,0 +1,48 @@
...
+
+     public function getCollection(string $resourceClass, string $operationName = null, array $context = []): iterable
+    {
+        Assert::keyExists($context, ContextKeys::CHANNEL);
+        Assert::keyExists($context, ContextKeys::LOCALE_CODE);
+
+        $channel = $context[ContextKeys::CHANNEL];
+        $localeCode = $context[ContextKeys::LOCALE_CODE];
+
+        return $this->productRepository->findLatestByChannel($channel, $localeCode, 3);
```
[the remark this answers is not in the document]
Comment (AdamKasp) [reply src/Sylius/Bundle/ApiBundle/DataProvider/LatestProductCollectionDataProvider.php]: I will improve it in the next iteration.

File: src/Sylius/Bundle/ApiBundle/spec/DataProvider/LatestProductCollectionDataProviderSpec.php [hunk 39]
```diff
@@ -0,0 +1,68 @@
...
+        ProductInterface $firstProduct,
+        ProductInterface $secondProduct
+    ): void {
+        $productRepository->findLatestByChannel($channel, 'en_US', 3)->willReturn([$firstProduct, $secondProduct]);
+
+        $this
+            ->getCollection(
+            ProductInterface::class,
+            'get_latest',
+            [ContextKeys::CHANNEL => $channel, ContextKeys::LOCALE_CODE => 'en_US'])
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/spec/DataProvider/LatestProductCollectionDataProviderSpec.php:64]: if the build will be green i will fix it in next PR

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductsContext.php [hunk 40]
```diff
@@ -288,6 +288,7 @@ public function iRemoveItsNameFromTranslation(string $localeCode): void
      */
     public function iShouldBeNotifiedThatItHasBeenSuccessfullyCreated(): void
     {
+        $test = $this->client->getLastResponse();
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductsContext.php:291]: nope

File: src/Sylius/Bundle/ApiBundle/spec/SerializerContextBuilder/ChannelContextBuilderSpec.php [hunk 41]
```diff
@@ -0,0 +1,31 @@
+<?php
+
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/SerializerContextBuilder/ChannelContextBuilderSpec.php:2]: missing license blocks and finals in specs

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/context_builders.xml [hunk 42]
```diff
@@ -0,0 +1,42 @@
...
+ For the full copyright and license information, please view the LICENSE
+ file that was distributed with this source code.
+
+-->
+
+<container xmlns="http://symfony.com/schema/dic/services"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd"
+>
+
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/context_builders.xml:18]: Unnecessary blank line

File: src/Sylius/Bundle/ApiBundle/spec/DataProvider/LatestProductCollectionDataProviderSpec.php [hunk 43]
```diff
@@ -0,0 +1,68 @@
...
+        $this->beConstructedWith($productRepository);
+    }
+
+    function it_supports_only_get_latest_operation_on_product_entity(): void
+    {
+        $this->supports(ProductInterface::class, 'get_latest')->shouldReturn(true);
+        $this->supports(TaxonInterface::class, 'get_latest')->shouldReturn(false);
+    }
+
+    function it_throw_exception_if_context_has_not_channel(): void
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/DataProvider/LatestProductCollectionDataProviderSpec.php:36]:
```suggestion
    function it_throws_an_exception_if_context_has_not_channel(): void
```

File: src/Sylius/Bundle/ApiBundle/spec/SerializerContextBuilder/ChannelContextBuilderSpec.php [hunk 44]
```diff
@@ -0,0 +1,40 @@
...
+        ChannelContextInterface $channelContext
+    ): void {
+        $this->beConstructedWith($decoratedContextBuilder, $channelContext);
+    }
+
+    function it_updates_context_when_channel_context_has_channel(
+        SerializerContextBuilderInterface $decoratedContextBuilder,
+        Request $request,
+        ChannelContextInterface $channelContext,
+        ChannelInterface $channel
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/SerializerContextBuilder/ChannelContextBuilderSpec.php:33]: As below

**Comment by GSadee** [comment]
Thanks, Adam! :tada:

**Pull request merged**